src/main/java/com/mparticle/kits/PilgrimSdkKit.java

```diff
@@ -0,0 +1,199 @@
+
+    /**
+     * Key used to get the Pilgrim Sdk api key from mParticle's settings response
+     */
+    static final String SDK_KEY = "pilgrim_sdk_key";
+
+    /**
+     * Key used to get the Pilgrim Sdk secret from mParticle's settings response
+     */
+    static final String SDK_SECRET = "pilgrim_sdk_secret";
+
+    /**
+     * Key used to get the Pilgrim Sdk enable  configuration enabling flag from mParticle's settings response
+     */
+    private static final String SDK_ENABLE_PERSISTENT_LOGS = "pilgrim_persistent_logs";
+
+    private static final String MPARTILE_USER_ID = "mParticleUserId";
+
+
+    @Override
+    protected List<ReportingMessage> onKitCreate(Map<String, String> settings, Context context) {
+        String key = settings.get(SDK_KEY);
+        if (KitUtils.isEmpty(key)) {
+            throw new IllegalArgumentException("PilgrimSdk key is empty.");
+        }
+
+        String secret = settings.get(SDK_SECRET);
+        if (KitUtils.isEmpty(secret)) {
+            throw new IllegalArgumentException("PilgrimSdk secret is empty.");
+        }
+
+        PilgrimSdk.Builder builder = new PilgrimSdk.Builder(context)
+                .consumer(key, secret)
+                .logLevel(LogLevel.DEBUG);
+
+        if (KitUtils.parseBooleanSetting(settings, SDK_ENABLE_PERSISTENT_LOGS, false)) {
+            builder.enableDebugLogs();
+        }
+
+        // Configure with our starter
+        PilgrimSdk.with(builder);
+
+        List<ReportingMessage> messageList = new LinkedList<>();
+        // Can we add messages to track if initialized/started successfully in attributes
+        messageList.add(new ReportingMessage(this, ReportingMessage.MessageType.APP_STATE_TRANSITION, System.currentTimeMillis(), null));
+        return messageList;
+    }
+
+    @Override
+    public String getName() {
+        return KIT_NAME;
+    }
+
+    @Override
+    public List<ReportingMessage> setOptOut(boolean optedOut) {
+        List<ReportingMessage> messageList = new LinkedList<>();
+        PilgrimSdk.stop(getContext());
+        messageList.add(new ReportingMessage(this, ReportingMessage.MessageType.OPT_OUT, System.currentTimeMillis(), null));
+        return messageList;
+    }
+
+    /**
+     * Returns ether last set user information, or a new
+     * PilgrimUserInfo instance.
+     */
+    private PilgrimUserInfo getUserInfo() {
+        PilgrimUserInfo info = PilgrimSdk.get().getUserInfo();
+        // can be null if it hasn't been set previously
+        if (info == null) {
+            info = new PilgrimUserInfo();
+        }
+        return info;
+    }
+
+    /* Section: User attribute set */
+    @Override
+    public void onSetAllUserAttributes(Map<String, String> attributes, Map<String, List<String>> attributeLists, FilteredMParticleUser filteredMParticleUser) {
+        // NOTE: attributeList not supported
+        // Is this ID correct?
+        updateUser(filteredMParticleUser.mpUser);
```

Comment:
In this case, you should pass the `filteredMParticleUser` instance itself into `updateUser()`. 

The `FilteredMParticleUser` implements `MParticleUser`, and it will reflect any filter settings a Client may have set up, whereas the underlying `MParticleUser` instance (`filteredMParticleUser.mpUser`), will not. It may be a misnomer that the method signature has `FilteredMParticleUser` rather than `MParticleUser` :)

Reply:
Updated!